### PR TITLE
[SPARK-32851][SQL][TEST] Tests should fail if errors happen when generating projection code

### PR DESCRIPTION
**Pull request by maropu**
### What changes were proposed in this pull request?

This PR intends to set `CODEGEN_ONLY` at `CODEGEN_FACTORY_MODE` in test spark context so that tests can fail if errors happen when generating expr code.

### Why are the changes needed?

I noticed that the code generation of `SafeProjection` failed in the existing test (https://issues.apache.org/jira/browse/SPARK-32828) but it passed because `FALLBACK` was set at `CODEGEN_FACTORY_MODE` (by default) in `SharedSparkSession`. To get aware of these failures quickly, I think its worth setting `CODEGEN_ONLY` at `CODEGEN_FACTORY_MODE`.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing tests.

### CODE REVIEW

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala [hunk 1]
```diff
@@ -104,6 +104,16 @@ class HiveQuerySuite extends HiveComparisonTest with SQLTestUtils with BeforeAnd
                                 | LIMIT 20
                               """.stripMargin
 
+  test("should fail if errors happen when generating expr code") {
```
Comment (HyukjinKwon) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala:107]: Shall we add a JIRA prefix?

Comment (maropu) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala:107]: Sure!

**Comment by maropu** [comment]
GA passed. cc: @cloud-fan @viirya

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 2]
```diff
@@ -2555,6 +2555,24 @@ class DataFrameSuite extends QueryTest
...
+  }
 }
 
 case class GroupByKey(a: Int, b: Int)
+
+case class FailedCodegenExpr(child: Expression) extends UnaryExpression with CodegenFallback {
+  private lazy val proj = FailedCodegenProjection.createObject(child :: Nil)
+  override def dataType: DataType = LongType
+  override def eval(input: InternalRow): Any = {
+    proj(input).getLong(0)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2576]: This test proves that if `eval` fails, the query fails. I think we should override `doCodeGen` and produce invalid code there?

Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2576]: This test proves that, if the codegen of `proj` fails, the query fails.  If `CODEGEN_FACTORY_MODE`=fallback, `proj` falls back into an interpreted one.

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2576]: Then shall we just add a test in `CodeGeneratorWithInterpretedFallbackSuite`,  to make sure that `FailedCodegenProjection` fails with default codegen mode?

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2576]: My point is that, the test here is not really an end-to-end test. It's just used to trigger `FailedCodegenProjection`

Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2576]: >  It's just used to trigger FailedCodegenProjection

Yea, yes. This test just triggers the the codegen failure of `FailedCodegenProjection` in executor sides and it checks the failure correctly propagates to a driver side in `SharedSparkSession` by setting the config `CODEGEN_FACTORY_MODE`=`codegen_only` in `SparkConf`:
https://github.com/apache/spark/pull/29721/files#diff-ef6933b96996929c402e3a36296b2590R71

`CodeGeneratorWithInterpretedFallbackSuite` is placed in the catalyst package, so it is not related to adding the config in `SharedSparkSession`. But, if the existing tests in  `CodeGeneratorWithInterpretedFallbackSuite` look enough, I think it is okay to drop the added tests and just add the config `CODEGEN_FACTORY_MODE` in `SharedSparkSession`.

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2576]: Yea I think it's OK to just change the config in test.

Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2576]: okay ~

**Review by cloud-fan — APPROVED** [review]

**Review by viirya — APPROVED** [review]

**Comment by maropu** [comment]
Thanks! Merged to master.

**Pull request closed without merging**